designs/function-node-lifecycle/README.md

```diff
@@ -41,17 +41,120 @@ Following figure shows new lifecycle model of function node.  While creating a f
+            ".*"              // disallow others 
+        ]
+    },
+```
+If the `externalModule` property not exists, `none` mode and denyList of `".*"`, meaning not allow installation and use of external module, is expected as a default.
+
+#### New API for installing NPM module
+
+In order to allow NPM modules to be installed automatically before node execution, `RED.nodes.registerType` will be extended to accept new `dynamicModuleList` property in an optional third parameter.  It may have following value:
+
+1.  a `string` value that represents a property name of a node instance.  The property of the node instance must have a list of modules to install.
+
+   ```
+   // Node definition
+   RED.nodes.registerType("function", FunctionNode, {
+       dynamicModuleList: "modules"
+   });
+   ```
+
+   ```
+   // Node instance definition
+   {
+       ...
+       "modules": [ "fs-ext@1.2.3", "qrcode@1.4.4" ]
+       ...
+   }
+   ```
+
+2. or an array of objects that must have a `name` property and may have a `scope` propety (but can have any other properties)
+
+   ```
+   // Node definition
+   RED.nodes.registerType("function", FunctionNode, {
+       dynamicModuleList: "modules"
+   });
+   ```
+
+   ```
+   // Node instance definition
+   {
+       ...
+       "modules": [ 
+           { 
+               "name": "fs-ext@1.2.3", 
+               "scope": "function:xyz",
+               "var": "fs", 
+               ... 
+           },
+           { 
+               "name": "qrcode@1.4.4", 
+               "scope": "function:xyz",
+               "var": "qrcode", 
+               ... 
+           }
+       ],
+       ...
+   }
+   ```
+
+   The Node-RED runtime will scan the list of modules to install in nodes deploy process before starting nodes execution.  A module in the list will be installed, if auto mode is specified in `settings.js` .  If `scope` is specified, NPM module is installed locally to specified scope.  On removal of a node with locally installed NPM modules, the node must uninstall the installed modules.
+
+   In order to support installation of NPM modules, the Node-RED runtime provides following API:
+
+   - Runtime API for Nodes
+
+     - `RED.nodes.addModule(opts)`
+     - `RED.nodes.removeModule(opts)`
+     - `RED.nodes.getModuleInfo(opts)`
+
+     Node API for nodes is extended to include these functions.  These new APIs accept the same parameter object `opts` to corresponding runtime APIs but works on generic NPM modules. 
+
+   - Editor API
+
+     Currently no editor APIs and endpoints are publicly provided for installing NPM module. But, following endpoints will be used for interaction between editor and runtime for fnction node:
+     
+     - `PUT /function/modules/` - install or update NPM module
+     - `GET /function/modules/` - get a list of installed NPM modules
+     - `GET /function/modules/:module` - get a NPM module's information
+     - `DELETE /function/modules/:module` - uninstall a NPM module
```

Comment:
As we're building this as a core feature of the runtime - supporting dynamic install of other modules, then I don't think these API endpoints should be under `/function`.

Is there a reason to not have them all be under `/modules/...`?

Reply:
At the time of writing this proposal, the endpoint was intended to be function node specific. Therefore, I added the `/function` prefix.　As you pointed out, it doesn't matter to remove the prefix.